packages/@lwc/integration-karma/test/static-content/index.spec.js

```diff
@@ -432,3 +434,51 @@ describe('static parts applies to comments correctly', () => {
         expect(refs.bar).toBe(bar);
     });
 });
+
+describe('static optimization with attributes', () => {
+    it('applies expressions on mount', async () => {
+        const elm = createElement('x-attribute-expression', { is: AttributeExpression });
+        document.body.appendChild(elm);
+        await Promise.resolve();
+
+        const nodes = extractDataIds(elm);
+        expect(nodes['loaded-on-mount'].getAttribute('data-mount')).toEqual('loaded on mount');
+        expect(nodes['loaded-on-mount-nested'].getAttribute('data-mount')).toEqual(
+            'loaded on mount'
+        );
+    });
+
+    it('updates expressions when data changes', async () => {
+        const elm = createElement('x-attribute-expression', { is: AttributeExpression });
+        document.body.appendChild(elm);
+        await Promise.resolve();
+
+        const nodes = extractDataIds(elm);
+        expect(nodes['loaded-on-update'].getAttribute('data-update')).toBe(null);
+        expect(nodes['loaded-on-update-nested'].getAttribute('data-update')).toBe(null);
+
+        elm.update = 'expression updated';
+        await Promise.resolve();
+
+        expect(nodes['loaded-on-update'].getAttribute('data-update')).toEqual('expression updated');
+        expect(nodes['loaded-on-update-nested'].getAttribute('data-update')).toEqual(
+            'expression updated'
+        );
+    });
+
+    it('applies expression from deeply nested data structure', async () => {
+        const elm = createElement('x-deep-attribute-expression', { is: DeepAttributeExpression });
+        document.body.appendChild(elm);
+        await Promise.resolve();
+
+        const nodes = extractDataIds(elm);
+        expect(nodes['deep1'].getAttribute('data-dynamic')).toEqual('1');
+        expect(nodes['deep1nested'].getAttribute('data-dynamic')).toEqual('1');
+        expect(nodes['deep2'].getAttribute('data-dynamic')).toEqual('2');
+        expect(nodes['deep2nested'].getAttribute('data-dynamic')).toEqual('2');
+        expect(nodes['deep3'].getAttribute('data-dynamic')).toEqual('3');
+        expect(nodes['deep3nested'].getAttribute('data-dynamic')).toEqual('3');
+        expect(nodes['deep4'].getAttribute('data-dynamic')).toEqual('4');
+        expect(nodes['deep4nested'].getAttribute('data-dynamic')).toEqual('4');
+    });
```

Comment:
It might be interesting to add a test with a variation on this template where everything is in one `<div>`. That way, they are all separate `part`s of a single static vnode.

Reply:
I'll make a note and include with the `style` and `class` attributes.